### ISSUE TITLE
`err` is sometimes a plain object

### ISSUE THREAD

**Issue by LinusU** [apiaryio/dredd#188]
In some cases the `err` parameter in, e.g, the callback to `dredd.run` isn't an error but instead an object.

This is visible in the code which features callback calls like this:

```coffee
callback({message: "Blueprint file or files not found on path: '#{config.options.path}'"}, stats)
```

This should instead be written like this:

```coffee
callback(new Error("Blueprint file or files not found on path: '#{config.options.path}'"), stats)
```

**Comment by netmilk** [comment]
Hi @LinusU, thank you very much for the report. It should be definitely consistent. I'll try to find out whether it's not an intention and fix it in the next release.

I'm curious how you do you use Dredd. What's your application for Dredd's JS API?

**Comment by LinusU** [comment]
I'm currently trying to run automated tests from the blueprint and stumbled on to this project which seem to have the same goal. I'm in the very early stages but currently I've added a new file to my `test/` directory which gets run with [mocha](https://github.com/mochajs/mocha).

I want to use it like this:

```javascript
/* eslint-env mocha */

var Dredd = require('dredd')

var dreddCfg = {
  path: ['Spec.apib']
}

describe('API', function () {
  var dredd

  before(function () {
    dredd = new Dredd(dreddCfg)
  })

  it('follows the blueprint', function (done) {
    dredd.run(done)
  })
})
```

This is how I noticed the problem since mocha outputed the following:

```text
  1) API follows the blueprint:
     Error: done() invoked with non-Error: {"message":"Blueprint file or files not found on path: ''"}
    at test/blueprint.js:19:23
    at Context.<anonymous> (test/blueprint.js:17:11)
```

It obviously also have some problem with locating my file but that's another issue altogether.

-----

What I really would want would be something like this:

```javascript
describe('API - Blueprint', function () {
  var tests = Dredd.loadSync(...)

  tests.forEach(function (test) {
    it(test.name, function (done) {
      test.run(done)
    })
  })
})
```

This would let mocha output the status of each individual test.

**Comment by netmilk** [comment]
Interesting, can I ask why do you prefer using Dredd JS API  and putting it into Mocha `it` block instead of use of CLI Dredd or [`grunt-dredd`](https://github.com/mfgea/grunt-dredd) wrapper to make a test of your blueprint part of the overall test suite and analyse unix exit status?

**Comment by LinusU** [comment]
You most certainly can :)

I'm using `mocha` for all my other tests and I do think that I will need to have manually specified tests in addition to api blueprint to catch all the edge cases. Therefor I wanted it to integrate nicely with the current test framework instead of being a different tool needed to be run.

I'm not using `grunt-dredd` since I'm not using `grunt` (currently for this project no build system at all but I usually use gulp), but the code in that module is very similar to the code that I have in my test-file.

Basically it boils down to keeping it in line with the rest of the tests. It will live under `test/` and it will be run as part of the test-suite at large when running `npm test`.

**Comment by LinusU** [comment]
@netmilk I've [modified protagonist](https://github.com/apiaryio/protagonist/pull/69) and are using that to give me the raw ast. It gives me the option to hook into mocha in a very nice way:

```javascript
describe('API', function () {
  var source = fs.readFileSync('Spec.apib', 'utf-8')
  var parsed = protagonist.parseSync(source)

  it.skip('should not emit any warnings', function () {
    assert.equal(parsed.warnings.length, 0)
  })

  parsed.ast.resourceGroups.forEach(function (resourceGroup) {
    describe(resourceGroup.name, function () {
      resourceGroup.resources.forEach(function (resouce) {
        describe(resouce.name, function () {
          resouce.actions.forEach(function (action) {
            it('should ' + action.name, function (done) {
              // TODO: do the actual test
              done()
            })
          })
        })
      })
    })
  })
})
```

```text
  API
    - should not emit any warnings
    People
      People List
        ✓ should List People
        ✓ should Create New Person
      Person
        ✓ should Get Person
        ✓ should Update Person
    Pets
      Pets List
        ✓ should List Pets
        ✓ should Create Pet
      Pet
        ✓ should Get Pet
        ✓ should Update Pet
        ✓ should Delete Pet
```

**Comment by netmilk** [comment]
@LinusU I'm not sure, if it's a lucky decision, because, you're actually reinventing the wheel which was already invented by writing Dredd. You will have to solve everything what's Dredd doing e.g. URI template expanding, URI parameters and headers inheriting, JSON schema generation, generating body, headers comparison etc..

**Comment by LinusU** [comment]
I agree, I hope to be able to use Dredd, but I'm just testing the water.

I really want to find the best way to integrate Dredd and Mocha, and I think that more people than me would benefit from that. Maybe it will be feasible to extract some part of dredd into `dredd-engine` or something similar, mainly the part that takes a part of an ast and generates a request, and the part that validates the response. Anyhow, just thinking out loud...

**Comment by netmilk** [comment]
Just letting know @LinusU [I'm going to extract](https://github.com/apiaryio/dredd/issues/226) this HTTP transaction compiler which you call `dredd-engine` soon for possible reuse in other projects

**Comment by honzajavorek** [comment]
The original subject of this issue was just fixed by #386.
